### ISSUE TITLE
Erzeugung von Kollisions-Objekten sinnvoll?

### ISSUE THREAD

**Issue by manuel3108** [manuel3108/Gleisbelegung#103]
Wie kann man das mit der Vorlaufzeit (beschrieben in #102 ) am besten Umsetzen.
Meine Idee wäre jetzt Kollisionsobjekte zu erstellen und in einer List zu speichern, und die von mir aus minütlich Abzufragen, ob davon etwas angezeigt werden soll.

Ansonsten könnten die Kollisionsobjekte auch einen Thread bekommen, und sich selbst um ihren Kram kümmern.

Was ist besser? Oder noch andere Ideen?

#100

**Comment by manuel3108** [comment]
Ok, dann wird das automatische vorschlagen von alternativ-Bahnsteigen auf jedem Fall nicht so einfach. Man könnte das aber dann so machen, dass man einfach alle als alternative-Bahnsteige gespeichert Bahnsteige anzeigt.

Zum speicher finde ich arrays nicht optimal, da ja nach einiger Zeit auch am Ende immer wieder neue TimeTabeleRow´s erstellt werden und man die Array´s dann immer neu erzeugen müsste. Das finde ich nicht sinnvoll. Auch kommt mir deine Speicherungsidee eher etwas kompliziert vor.
Ich würde eher so etwas in dieser Richtung machen.
```
class Kollision{
  List<TimeTable.TimeTableData> doppeltBelegung;     //die Element die eine doppelBelegung kennzeichnen 
                           (müüsen hintereinander und im selben Bahnsteig sein)
  Bahnsteig b; //muss nocht sein, kriegt man aus TTData und TTColumn
  
  void ploppeAuf();
  void keineAhnung();
}
```

Ich würde das ganze dann als `List<Kollision> kollisionen` der Klasse Gleisbelegung implementieren. Das ganze dann über update() minütlich abfragen und dann zunächst anzeigen. 
Was meinst du dazu?

**Comment by Greonyral** [comment]
Zu List statt array: Du benutzt als List, in der Regel ArrayList, was im wesentlichen ein array in einer Klasse mit ein paar Methoden ist. Das schliesst sich quasi nicht gegenseitig aus ;).
Es werden neue Rows erstellt ja, Aber die Gesamtzahl der Reihen ist höchtens die Einstellung der Minuten, wie weit das Plugin in die Zukunft blicken soll. Die x.-te Reihe ist des Arrays bildet damit die Reihe nach jetzt + x Minuten ab. Daher würde die Verknüpfung mit einer speziellen TimeTableRow erst später machen (zB. als Wert des vorgeschlagenen 2D-Arrays).
Bei deinem Vorschlag fehlt das "Wie findet da Plugin raus, dass eine Doppelbelgung stattfindet?" Dein Vorschlag bezieht sich im Wesentlichen nur auf "Wie speicher ich das Ergebnis?"

> Ok, dann wird das automatische vorschlagen von alternativ-Bahnsteigen auf jedem Fall nicht so einfach. Man könnte das aber dann so machen, dass man einfach alle als alternative-Bahnsteige gespeichert Bahnsteige anzeigt.

Mein Vorschlag wäre 2-stufig, erst alle während des Konflikts immer freien Nachbarbahnsteige (Ausweichmöglichkeit schafft keine neuen Konflikte), und wenn da nichts zu finden ist, die gesamte Liste anzuzeigen

**Comment by manuel3108** [comment]
> Zu List statt array: Du benutzt als List, in der Regel ArrayList, was im wesentlichen ein array in einer Klasse mit ein paar Methoden ist. Das schliesst sich quasi nicht gegenseitig aus ;).

Ja, gerade heute in der Vorlesung gelernt. Hätte man sich aus dem namen ja auch schließen können, aber gut.

> Es werden neue Rows erstellt ja, Aber die Gesamtzahl der Reihen ist höchtens die Einstellung der Minuten, [...]

Das ist tatsächlich mal so gewesen, in der aktuellen nur für uns hier verfügbaren Version ist dies allerdings nicht mehr der Fall. Dies ist in älteren Versionen auch nur der Fall gewesen, da das nicht alles so dynamisch wie jetzt generiert wurde. So wie es momentan ist, werden alle erstellten TTRows auch angezeigt. Dadurch würde ich dann aber tatsächlich zu einer List tendieren.

> Mein Vorschlag wäre 2-stufig, erst alle während des Konflikts immer freien Nachbarbahnsteige (Ausweichmöglichkeit schafft keine neuen Konflikte), und wenn da nichts zu finden ist, die gesamte Liste anzuzeigen

Ja, das klingt sinnvoll und auch für den Benutzer sehr angenehm. Ich könnte mir allerdings vorstellen, dass das manchmal für Verwirrung sorgt, so nach dem Motto `"Warum wird in Situation x auf Gleis z nur nur Gleis a angezeigt"` und `"Warum wird in Situation y auf Gleis z nur Gleis a und b angezeigt"`.
Deshalb würde ich alle Bahnsteige anzeigen, die ohne Kollision aber zum Beispiel in grün und die anderen in rot oder so. Was meinst du?

> Bei deinem Vorschlag fehlt das "Wie findet da Plugin raus, dass eine Doppelbelgung stattfindet?" Dein Vorschlag bezieht sich im Wesentlichen nur auf "Wie speicher ich das Ergebnis?"

Du hast natürlich recht, das scheint mir da untergegangen zu sein. Im Grunde genommen stimme ich dir aber mit deiner Idee Vollkommen überein:
> Ist die Größe einer Liste größer 1, ist eine Kolllision vorhanden, freie Bahnsteige sind alle Bahnsteige die während der Haltezeit eines Zuges kontinuierlich leer ist. IM Beispiel wäre E doppelt belegt, und zwar 1-4 und 3-5 bzw 1-5 und 3-4. Im ersten Fall gibts für den zweiten Zug auf A eine Alternativ, im zweiten Fall auf A und D.

Für mich ist an der Stelle allerdings nur ein bisschen Fraglich wo man das genau einbaut. Am einfachsten wäre es natürlich in TimeTable, aber da hätten wir wieder das Thema, dass wir eigentlich nicht zu sehr vom Sinn der Klasse abweichen wollen. Dennoch habe ich momentan keine andere Idee, an der es auch inhaltlich passen würde.
